ext/standard/string.c

```diff
@@ -1403,13 +1406,133 @@ PHPAPI char *php_strtoupper(char *s, size_t len)
 }
 /* }}} */
 
+#if !defined(ZTS) && defined(__aarch64__)
+static zend_always_inline void memcpy_15(uint8_t *out, uint8_t *in, uint8_t len) {
+	uint8_t k = 0;
+	if (len >= 8) {
+		memcpy(out, in, 8);
+		len -= 8;
+		k += 8;
+	}
+	if (len >= 4) {
+		memcpy(out + k, in + k, 4);
+		len -= 4;
+		k += 4;
+	}
+	if (len >= 2) {
+		memcpy(out + k, in + k, 2);
+		len -= 2;
+		k += 2;
+	}
+	if (len == 1) {
+		memcpy(out + k, in + k, 1);
+	}
+}
+
+static zend_always_inline uint8x16_t neon_to_lu(uint8x16_t x,
+						uint8_t shift_LUT[4][64],
+						uint8_t has_shift[4],
+						uint8x16_t *transformed) {
+	uint8x16_t shifts = {0};
+
+	/* There are no lower/upper case chars in the first table: [0, 63]. */
+	int i = 1;
+	for (; i < 4; i++) {
+		if (!has_shift[i])
+			continue;
+
+		/* Mask out the higher bits that are used to index in the
+		   shift_LUT[i]: a vector table lookup that accesses in the
+		   range [0..63] will return the value recorded in the LUT,
+		   otherwise the returned value is 0. The value in the LUT is
+		   the shift amount to convert low->up or up->low case. */
+		uint8x16_t idx = veorq_u8(x, vdupq_n_u8(i << 6));
+		uint8x16_t s = vqtbl4q_u8(*((uint8x16x4_t *)shift_LUT[i]), idx);
+		shifts = vaddq_u8(shifts, s);
+	}
+
+	/* The output is transformed if any char is shifted. */
+	*transformed = vorrq_u8(*transformed, shifts);
+	return vaddq_u8(x, shifts);
+}
+
+static zend_always_inline
+zend_string *neon_string_to_lu(zend_string *s,
+			       uint8_t shift_LUT[4][64],
+			       uint8_t has_shift[4])
+{
+	size_t len = ZSTR_LEN(s);
+	zend_string *res = zend_string_alloc(len, 0);
```

Comment:
This implementation always duplicates the original string (even, if it is already in lower case). This may cause degradation.

Reply:
Please correct me if I am wrong: my understanding is that the concern is about allocating and free-ing too much memory.  The case where the string is already lower case is taken care of to not duplicate the memory in the current version of the patch, however the current implementation still needs a separate buffer that gets allocated/freed in a short amount of time. 

I have implemented a version that avoids the short alloc/free: https://github.com/php/php-src/commit/0fc99c5f556692d75454f0d6841fbf6db75a19ba
and the benchmarks for tolower on an all lower case string show that it is slower than the current version on review.  The reason is that the new implementation requires control flow dependent on the processed data: the transfer between NEON and general purpose registers is 2 cycles and that is required to do the branch in the tight loop on whether we have seen an upper case so far or not.  The current patch under review avoids this check by duplicating the data and only at the very end of the loop it does the transfer from NEON register to the GPR.
The benchmarks that I was looking at are showing that more short-lived alloc/free do not impact the performance as much as the NEON->GPR transfers.

Do you have a benchmark I should look at for performance analysis that is more relevant than the one I described in the patch commit message?  Thanks!